documentation/src/main/asciidoc/backend-lucene.asciidoc

```diff
@@ -14,14 +12,184 @@ but the defaults might not suit everyone.
 In particular, you might want to set the location of your indexes in the filesystem.
 See below for the details of every configuration property.
 
-[[backend-lucene-configuration-directory-provider]]
-=== Directory provider
+[[backend-lucene-configuration-directory]]
+=== Index storage (`Directory`)
 // Search 5 anchors backward compatibility
 [[search-configuration-directory]]
 
-include::todo-placeholder.asciidoc[]
+The component responsible for index storage in Lucene is the `org.apache.lucene.store.Directory`.
+The implementation of the directory determines where the index will be stored:
+on the filesystem, in the JVM's heap, ...
+
+By default, the Lucene backend stores the indexes on the filesystem,
+in the JVM's working directory.
+
+The type of directory is set a the backend level:
+
+// Search 5 anchors backward compatibility
+[[example-configuring-directory-providers]]
+[source]
+----
+hibernate.search.backends.<backend-name>.directory.type = local-filesystem
+----
+
+// Search 5 anchors backward compatibility
+[[directory-provider-table]]
+The following directory types are available:
+
+* `local-filesystem`: Store the index on the local filesystem.
+See <<backend-lucene-configuration-directory-local-filesystem>>
+for details and configuration options.
+* `local-heap`: Store the index in the local JVM heap.
+*Local heap directories and all contained indexes are lost when the JVM shuts down.*
+See <<backend-lucene-configuration-directory-local-heap>>
+for details and configuration options.
+
+[[backend-lucene-configuration-directory-local-filesystem]]
+==== Local filesystem storage
+
+The `local-filesystem` directory type will store each index in
+a subdirectory of a configured filesystem directory.
+
+[NOTE]
+====
+Local filesystem directories really are designed to be *local* to one server and one application.
+
+In particular, they should not be shared between multiple Hibernate Search instances
+(example: network share).
+
+If you need to share indexes between multiple Hibernate Search instances,
+the Elasticsearch backend will be a better choice.
```

Comment:
Is the Elasticsearch backend the only option to have a working cluster of Hibernate Search instances? What about Infinispan directory provider? Maybe we could add a `so far`...

Reply:
The Infinispan directory provider 1) does not exist for Search 6 yet and 2) is not enough to cluster Hibernate Search, since we also need to route reindexing events from slave nodes to the master node.

Support for clustering is not planned until 6.1, and even then I'm not sure we will address it for the Lucene backend, just the Elasticsearch one, at least for a start. We may extend that to Lucene if it's really easy, but we just don't have the manpower necessary to maintain our own clustered search engine solution.

Infinispan query is another story entirely, since they handle clustering themselves.